File: tools/LabKeyEncrypter/LabKeyEncrypter.Library/LabKeyJsonFileValueEncrypter.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace LabKeyEncrypter.Library;

/// <summary>
/// Encrypts and decrypts JSON file values using AES-256-GCM.
/// </summary>
public static class LabKeyJsonFileValueEncrypter
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        WriteIndented = true
    };

    /// <summary>
    /// Encrypts all top-level values in a JSON file and saves to a new file.
    /// </summary>
    /// <param name="filePath">Path to the JSON file to encrypt.</param>
    /// <param name="password">Password for encryption.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    public static async Task EncryptJsonValuesAsync(string filePath, string password, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(filePath);
        ArgumentException.ThrowIfNullOrEmpty(password);

        var jsonString = await File.ReadAllTextAsync(filePath, cancellationToken);
        var jsonObject = JsonNode.Parse(jsonString)?.AsObject()
            ?? throw new JsonException("Invalid JSON: expected an object");

        foreach (var (key, value) in jsonObject.ToArray())
        {
            var plainText = value?.ToJsonString() ?? "null";
            jsonObject[key] = LabKeyEncrypter.Encrypt(plainText, password);
        }

        var encryptedFilePath = Path.ChangeExtension(filePath, null) + "_encrypted.json";
        await File.WriteAllTextAsync(encryptedFilePath, jsonObject.ToJsonString(JsonOptions), cancellationToken);
    }

    /// <summary>
    /// Encrypts all top-level values in a JSON file and saves to a new file (synchronous).
    /// </summary>
    public static void EncryptJsonValues(string filePath, string password) =>
        EncryptJsonValuesAsync(filePath, password).GetAwaiter().GetResult();

    /// <summary>
    /// Decrypts all top-level values in an encrypted JSON file and saves to a new file.
    /// </summary>
    /// <param name="filePath">Path to the encrypted JSON file.</param>
    /// <param name="password">Password for decryption.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>True if decryption was successful.</returns>
    public static async Task<bool> DecryptJsonValuesAsync(string filePath, string password, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(filePath);
        ArgumentException.ThrowIfNullOrEmpty(password);

        var jsonString = await File.ReadAllTextAsync(filePath, cancellationToken);
        var jsonObject = JsonNode.Parse(jsonString)?.AsObject()
            ?? throw new JsonException("Invalid JSON: expected an object");

        foreach (var (key, value) in jsonObject.ToArray())
        {
            var encrypted = value?.GetValue<string>()
                ?? throw new JsonException($"Expected string value for key '{key}'");
            var decrypted = LabKeyEncrypter.Decrypt(encrypted, password);
            jsonObject[key] = JsonNode.Parse(decrypted);
        }

        var decryptedFilePath = filePath.Replace("_encrypted.json", ".json", StringComparison.OrdinalIgnoreCase);
        await File.WriteAllTextAsync(decryptedFilePath, jsonObject.ToJsonString(JsonOptions), cancellationToken);
        return true;
    }

    /// <summary>
    /// Decrypts all top-level values in an encrypted JSON file (synchronous).
    /// </summary>
    public static bool DecryptJsonValues(string filePath, string password) =>
        DecryptJsonValuesAsync(filePath, password).GetAwaiter().GetResult();
}

File: tools/LabKeyEncrypter/LabKeyEncrypter.Python/lab_key_json_file_value_encrypter.py
"""
LabKeyJsonFileValueEncrypter - Encrypt/decrypt JSON file values.

Compatible with the C# LabKeyJsonFileValueEncrypter implementation.
"""

import json
from pathlib import Path
from lab_key_encrypter import encrypt, decrypt


def encrypt_json_values(file_path: str, password: str) -> None:
    """
    Encrypt all values in a JSON file and save to a new file.
    
    Args:
        file_path: Path to the JSON file to encrypt.
        password: The password to use for encryption.
    """
    path = Path(file_path)
    json_string = path.read_text(encoding='utf-8')
    json_object = json.loads(json_string)
    
    encrypted_object = {}
    for key, value in json_object.items():
        plain_text = json.dumps(value) if not isinstance(value, str) else value
        encrypted_object[key] = encrypt(plain_text, password)
    
    encrypted_file_path = str(path).replace('.json', '_encrypted.json')
    Path(encrypted_file_path).write_text(
        json.dumps(encrypted_object, indent=2),
        encoding='utf-8'
    )


def decrypt_json_values(file_path: str, password: str) -> bool:
    """
    Decrypt all values in an encrypted JSON file and save to a new file.
    
    Args:
        file_path: Path to the encrypted JSON file.
        password: The password used for encryption.
    
    Returns:
        True if decryption was successful.
    """
    path = Path(file_path)
    json_string = path.read_text(encoding='utf-8')
    json_object = json.loads(json_string)
    
    decrypted_object = {}
    for key, value in json_object.items():
        decrypted_text = decrypt(value, password)
        # Try to parse as JSON, otherwise use as string
        try:
            decrypted_object[key] = json.loads(decrypted_text)
        except json.JSONDecodeError:
            decrypted_object[key] = decrypted_text
    
    decrypted_file_path = str(path).replace('_encrypted.json', '.json')
    Path(decrypted_file_path).write_text(
        json.dumps(decrypted_object, indent=2),
        encoding='utf-8'
    )
    
    return True
